File: core/c/src/rng.c
/* VOIDSTRIKE — Protocol v1 · splitmix64 (PROTOCOL §4). */
#include "voidstrike/rng.h"

uint64_t vs_rng_next(uint64_t *state)
{
    uint64_t z;
    *state += 0x9E3779B97F4A7C15ULL;
    z = *state;
    z = (z ^ (z >> 30)) * 0xBF58476D1CE4E5B9ULL;
    z = (z ^ (z >> 27)) * 0x94D049BB133111EBULL;
    return z ^ (z >> 27);
}

double vs_rng_uniform(uint64_t *state)
{
    return (double)vs_rng_next(state) / 18446744073709551616.0; /* 2^64 */
}

File: src/lib/sim/rng.ts
/**
 * VOIDSTRIKE — splitmix64 RNG (PROTOCOL.md §4).
 * Single shared stream per world. Exact u64 arithmetic via BigInt so
 * results are bit-identical to the C reference core on every platform.
 *
 * uniform01(): next() / 2^64  — Number(bigint) is the correctly-rounded
 * double of the integer; dividing by the exact power of two 2^64 is a pure
 * exponent scaling, so the result is deterministic across engines.
 */

const GOLDEN = 0x9e3779b97f4a7c15n;
const M1 = 0xbf58476d1ce4e5b9n;
const M2 = 0x94d049bb133111ebn;
const U64_MASK = 0xffffffffffffffffn;
const TWO_POW_64 = 18446744073709551616.0; // 2^64 exactly representable

export class SplitMix64 {
  private state: bigint;

  constructor(seed: number | bigint) {
    this.state = BigInt(seed) & U64_MASK;
  }

  /** Raw u64 output. */
  nextU64(): bigint {
    this.state = (this.state + GOLDEN) & U64_MASK;
    let z = this.state;
    z = ((z ^ (z >> 30n)) * M1) & U64_MASK;
    z = ((z ^ (z >> 27n)) * M2) & U64_MASK;
    return (z ^ (z >> 27n)) & U64_MASK;
  }

  /** Uniform double in [0, 1). */
  uniform01(): number {
    return Number(this.nextU64()) / TWO_POW_64;
  }

  /** Deterministic snapshot of the stream state (for replays / debug). */
  getState(): string {
    return this.state.toString(16).padStart(16, "0");
  }
}
